### ISSUE TITLE
[Datasets] Actors are incorrectly cached in local mode

### ISSUE THREAD

**Issue by anaoum** [ray-project/ray#26389]
### What happened + What you expected to happen

When running an operation on a dataset (such as map, or map_batches), and using an actor strategy by passing in a CallableClass as the `fn`, and `compute="actors"`, the first actor class that is used will be cached and re-used, even if subsequent calls specify a new actor class. This only seems to be the case when ray has been initialised with `local_mode=True`.

### Versions / Dependencies

ray==1.13.0
Python 3.8.13

### Reproduction script

```python
import ray

class Actor1:
    def __call__(self, item):
        return "a1"

class Actor2:
    def __call__(self, item):
        return "a2"

ray.init(num_cpus=1, local_mode=True)

data = ray.data.from_items(["a"]).map(Actor1, compute="actors").take_all(1)
assert data == ["a1"], f"expected ['a1'], not {data}"

data = ray.data.from_items(["a"]).map(Actor2, compute="actors").take_all(1)
assert data == ["a2"], f"expected ['a2'], not {data}"
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by clarkzinzow** [comment]
@anaoum Can you try this on latest master? I'm not able to reproduce this on latest master, I think this might have been fixed incidentally in this commit: https://github.com/ray-project/ray/commit/caa3868570cfb39500766cdddf956b8494511e6b

**Comment by anaoum** [comment]
I can confirm that the issue is:

- resolved in d234348bd2cd0a6cf1d300c228e07980e4053268 (latest master)
- resolved in caa3868570cfb39500766cdddf956b8494511e6b (the commit you mentioned)
- present in 647f6790c07bfe5677fc0598772dace8c62fccbc (parent of the commit you mentioned)
